### PR TITLE
replace manage facility admin screen dropdowns with combos

### PR DESCRIPTION
**Pull request by fzhao99**
# FRONTEND PULL REQUEST

## Related Issue

Part one of #6393 

Will follow up with the unarchive page in a separate PR.

## Changes Proposed

- Refactor ManageFacility component to use Truss combo boxes rather than our custom dropdown
- Update tests accordingly
- Added a new custom `ComboBox` based on the Truss version with an added `required` prop

## Testing 

deployed on dev5

### Before
Screenshot 2023-08-31 at 10 04 29 AM

### After
https://github.com/CDCgov/prime-simplereport/assets/29645040/5c47f1f7-7044-42b1-983e-de77c5ac314a

### Org selection enables facility selection
https://github.com/CDCgov/prime-simplereport/assets/29645040/97a56504-cfff-4f54-98b5-750d91bbcb7d

### Clear filters on individual combos cascade
https://github.com/CDCgov/prime-simplereport/assets/29645040/205b392e-53ca-48a0-b944-454c798be851

### Clear filter button clears both boxes
https://github.com/CDCgov/prime-simplereport/assets/29645040/450e6533-7931-4b37-b903-14a2e2f9f534

### Clear org resets search results 
https://github.com/CDCgov/prime-simplereport/assets/29645040/be723f5f-9f31-4327-885a-3144ab02a69d

## Additional Information

TL;DR: Are we ok sticking with fireEvent / wrapping things in `act` like we're doing currently, or do we think it's worth it to try and prefer `userEvent` / clean `act` warnings in the future? The issue is that using fireEvent / suppressing act warnings with the extra tag [may result in tests that don't catch the behavior they were written for](https://egghead.io/lessons/jest-fix-the-not-wrapped-in-act-warning), as well as resulting in more stable tests because we're accounting for all the state updates that are happening within our component (to the extent React can tell us). On the flip side though, going back and refactoring might be expensive / unfun. I reworked some of the tests in the above fashion [here](https://github.com/CDCgov/prime-simplereport/pull/6461/files) in case more concrete examples are helpful.

If yes, happy to make a ticket and prioritize accordingly alongside the other tech debt work we have in the pipeline. Otherwise, we can revisit if it ever comes up again.

### The issue

In getting this PR in I discovered a couple of issues with the `act` warning React is throwing based on the way that we're setting up our testing suite. I'll outline what I've found here and would love opinions on options to move forward!

As we've addressed in PR's like [this one](https://github.com/CDCgov/prime-simplereport/pull/5950), `act` warnings pop up periodically in our component testing. [These are often useful warnings React offers to notify us about unhandled state changes](https://codilime.com/blog/why-should-you-be-grateful-for-act-warnings-in-react-tests/) within our components that might cause unintended side effects. 

For this PR specifically, we need to use the `userEvent` API rather than the lower-level `fireEvent` to get tests to pass because the combo box is listening for a couple different events to happen in sequence in order to update. This caused a bunch of `act` warnings to pop up and I wanted to present a few options of handling them so we don't unnecessarily clutter our app logs and write more robust tests.

#### Option 0 - Use fireEvent
For the reasons stated above, we can't exactly use fireEvent here because the component we're testing for relies on a bunch of bundled events working together, but we do use this method extensively in other parts of the codebase. [RTL specifically recommends against this option though](https://ph-fritsche.github.io/blog/post/why-userevent), so wanted to call it out as something to potentially move away from so we're testing things in a way more in line with how they [recommend](https://testing-library.com/docs/user-event/intro/#differences-from-fireevent). 

#### Option 1 - Wrap the userEvents in `act`
One way of suppressing the warnings is to wrap the events in an `act` call. [The RTL maintainer again recommends against this](https://kentcdodds.com/blog/fix-the-not-wrapped-in-act-warning#the-dreaded-act-warning) because the API calls already have `act` wrappers in them and the warnings might be pointing to actual state updates that need addressing. Back when Nick was on the project [he also submitted a PR that called this out](https://github.com/CDCgov/prime-simplereport/pull/3028) and installed [an eslint plugin specifically for this issue](https://github.com/testing-library/eslint-plugin-testing-library/blob/main/docs/rules/no-unnecessary-act.md). The plugin seems not to be working correctly because it's not yelling at us in instances where it should be.

I've opted to temporarily go this direction to get this PR in, but will mention that this is a bandaid solution since (most) of the `act` warnings are trying to tell us that we have side effects in our component code that could result in more brittle tests.

#### Option 2 - Update dependencies that are causing the issue in the first place and refactor our testing suite
This is, at least in my mind, the "correct" solution, but also the one that will take a bit of work. By using the higher-level interaction API _and_  making sure we're handling all the `act` warnings, we're ensuring that we're covering any side effects that happen within our event handlers / component rendering logic. [This video](https://egghead.io/lessons/jest-fix-the-not-wrapped-in-act-warning) outlines a scenario where the act warning is trying to tell you something useful.

In order for us to  to take advantage of `userEvent` without act warnings though, we need to make sure that the underlying [shared `@testing-library/dom` dependency resolves to the same version.](https://github.com/testing-library/user-event/issues/1104#issuecomment-1460005394). I started exploring this on a separate branch but updating the packages breaks some other tests, so I punted on this for now. If we're in agreement that we want to do this, we'll need to go back and update some old tests so that everything is up to standard (and fix the eslint plugin in the meantime so that it's enforcing the new set of API's).

### CODE REVIEW

File: cypress/e2e/11-support-manage-facility.cy.js [hunk 1]
```diff
@@ -38,11 +40,25 @@ describe("Support admin: manage facility", () => {
...
     cy.wait("@GetAllOrganizations");
-    cy.get("div.bg-base-lightest select").first().select(organizationId);
 
-    // selects facility
+    // selects org combo box
+    cy.get("[data-testid=\"org-selection-container\"] " +
+      "> [data-testid=\"combo-box\"] ")
+      .within(() => {
+        // within the org selection box
+        cy.get("[data-testid=\"combo-box-input\"]").click();
```
Comment (fzhao99) [inline cypress/e2e/11-support-manage-facility.cy.js:49]: I thought it was a little gross to use content-unaware selectors in our tests like this but couldn't get the selectors to work because of the way `select` within the combo box is hidden from view on default (as Cypress complains about the `select` not being in view if you try to select it). If anyone has suggestions on improving the way we're choosing selectors, would love that!

File: cypress/e2e/11-support-manage-facility.cy.js [hunk 2]
```diff
@@ -1,29 +1,31 @@
...
       facilityCreated.id = response.body.data.addFacility.id;
     });
 
     whoAmI().then((res) => {
       organizationId = res.body.data.whoami.organization.id;
       facilityId = res.body.data.whoami.organization.facilities[0].id;
     });
+  });
 
+  beforeEach(() => {
```
Comment (fzhao99) [inline cypress/e2e/11-support-manage-facility.cy.js:26]: moving this hook into a before each as recommended [here](https://docs.cypress.io/guides/core-concepts/variables-and-aliases#Aliases-are-reset-before-each-test)

File: cypress/e2e/11-support-manage-facility.cy.js (hunk 1 above)
Comment (emyl3) [reply cypress/e2e/11-support-manage-facility.cy.js:49]: for this the other alternative I could think of if you had both the org name and facility name handy is...
`cy.get('input[role="combobox"]').first().type("${ORG_NAME}{enter}");`
`cy.get('input[role="combobox"]').last().type("${FACILITY_NAME}{enter}");`

Comment (fzhao99) [reply cypress/e2e/11-support-manage-facility.cy.js:49]: thanks for the suggestions elisa! added.

**Comment by fzhao99** [comment]
Spike ticket for the act pieces made [here](https://github.com/CDCgov/prime-simplereport/issues/6491)

**Review by emyl3 — DISMISSED** [review]
LGTM! Thank you for doing this (combo boxes are so tricky 😭) and doing the initial research into the act warnings!!!

File: frontend/src/app/supportAdmin/ManageFacility/FacilitySelectFilter.tsx [hunk 3]
```diff
@@ -58,40 +64,47 @@ const FacilitySelectFilter: React.FC<FacilitySelectFilterProps> = ({
...
+            </Label>
+            <ComboBox
+              name={"Organization"}
+              id={"manage-facility-org-select"}
+              options={organizationOptions}
+              onChange={(val) => {
+                onSelectOrg(val);
+              }}
               disabled={loading}
-              required={true}
```
Comment (johanna-skylight) [inline frontend/src/app/supportAdmin/ManageFacility/FacilitySelectFilter.tsx:74]: The required true was needed to pass a11y scan. Does the combobox handle the required in a different way?

File: frontend/src/app/supportAdmin/ManageFacility/ManageFacility.tsx [hunk 4]
```diff
@@ -141,23 +161,26 @@ const ManageFacility = () => {
...
-    deleteFacilityMutation({
-      variables: { facilityId: localState.facilityId },
-    }).then(() => {
-      showSuccess(
-        "",
-        `Facility ${localState.facility?.name} successfully deleted`
-      );
-      handleClearFilter();
-    });
+    if (localState.facilityId) {
```
Comment (johanna-skylight) [inline frontend/src/app/supportAdmin/ManageFacility/ManageFacility.tsx:173]: why is this condition needed? in theory the delete button will display only if a facility has been loaded

**Review by johanna-skylight — CHANGES_REQUESTED** [review]
I left some questions/comments. Also I was not able to see the code to migrate the unarchive patient page. Is that going to be addressed in another PR or was it already migrated?.

Thanks for your work on integrating the trusswork component and sorry for the delayed review!

File: frontend/src/app/supportAdmin/ManageFacility/FacilitySelectFilter.tsx (hunk 3 above)
Comment (johanna-skylight) [reply frontend/src/app/supportAdmin/ManageFacility/FacilitySelectFilter.tsx:74]: @BobanL do you remember how you found the issue [6408](https://github.com/CDCgov/prime-simplereport/issues/6408)?. Want to make sure we don't need the required at the input level before resolving this combo! :D

Comment (BobanL) [reply frontend/src/app/supportAdmin/ManageFacility/FacilitySelectFilter.tsx:74]: This happens when you test this as a form. I ran through it and it looks like it's still being flagged. I think for the purpose of accessibility we might need to figure out how to add [required](https://www.deque.com/blog/anatomy-of-accessible-forms-required-form-fields/#:~:text=While%20all%20screen,can%20perceive%20it.).
image

Comment (fzhao99) [reply frontend/src/app/supportAdmin/ManageFacility/FacilitySelectFilter.tsx:74]: oh actually Truss gives us an escape hatch for this. Will push a fix soon

Comment (fzhao99) [reply frontend/src/app/supportAdmin/ManageFacility/FacilitySelectFilter.tsx:74]: Screenshot 2023-09-06 at 4 35 30 PM

The latest commit adds to prop but if someone could run the scan for me again would appreciate it!

Comment (BobanL) [reply frontend/src/app/supportAdmin/ManageFacility/FacilitySelectFilter.tsx:74]: Tested it with axe and didn't find any problems!

**Review by BobanL — APPROVED** [review]
Looks good to me!

File: frontend/src/app/supportAdmin/ManageFacility/FacilitySelectFilter.tsx (hunk 3 above)
Comment (emyl3) [reply frontend/src/app/supportAdmin/ManageFacility/FacilitySelectFilter.tsx:74]: When I ran it on the latest changes I saw this error but am ignoring because I believe this is a false positive jsyk
Screenshot 2023-09-07 at 09 42 39

**Review by emyl3 — APPROVED** [review]
Thank you for all your work on these tricky tricky combo boxes 🎃

**Review by johanna-skylight — APPROVED** [review]
👍
